analytic_engine/src/instance/flush_compaction.rs

```diff
@@ -223,74 +219,20 @@ impl Flusher {
             table_data, opts
         );
 
-        let flush_req = self.preprocess_flush(table_data).await?;
-
-        self.schedule_table_flush(flush_scheduler, flush_req, opts, true)
+        self.schedule_table_flush(flush_scheduler, table_data.clone(), opts, true)
             .await
     }
 
-    async fn preprocess_flush(&self, table_data: &TableDataRef) -> Result<TableFlushRequest> {
-        let current_version = table_data.current_version();
-        let last_sequence = table_data.last_sequence();
-        // Switch (freeze) all mutable memtables. And update segment duration if
-        // suggestion is returned.
-        if let Some(suggest_segment_duration) =
-            current_version.switch_memtables_or_suggest_duration()
-        {
-            info!(
-                "Update segment duration, table:{}, table_id:{}, segment_duration:{:?}",
-                table_data.name, table_data.id, suggest_segment_duration
-            );
-            assert!(!suggest_segment_duration.is_zero());
-
-            let mut new_table_opts = (*table_data.table_options()).clone();
-            new_table_opts.segment_duration = Some(ReadableDuration(suggest_segment_duration));
-
-            let edit_req = {
-                let meta_update = MetaUpdate::AlterOptions(AlterOptionsMeta {
-                    space_id: table_data.space_id,
-                    table_id: table_data.id,
-                    options: new_table_opts.clone(),
-                });
-                MetaEditRequest {
-                    shard_info: table_data.shard_info,
-                    meta_edit: MetaEdit::Update(meta_update),
-                }
-            };
-            self.space_store
-                .manifest
-                .apply_edit(edit_req)
-                .await
-                .context(StoreVersionEdit)?;
-
-            // Now the segment duration is applied, we can stop sampling and freeze the
-            // sampling memtable.
-            current_version.freeze_sampling();
-        }
```

Comment:
I guess we can't do manifest update without holding the `serial_exec`, that is to say, storing the sampled segment duration into Manifest should still be processed here. However, we can freeze normal memtable in the flush task.

So my suggestion is to split the preprocess_flush into two functions:
- maybe_update_smapled_segment_duration: call it before schedule a flush task
- freeze_memtable: call it in the flush task

Reply:
I'm wrong, this is ok, there is no need to hold the `serial_exec` for manifest update.